packages/controllers/src/snaps/RequestQueue.ts

```diff
@@ -0,0 +1,59 @@
+import { ReentrancyGuard } from './ReentrancyGuard';
+
+export class RequestQueue {
+  public readonly maxQueue: number;
+
+  private readonly locks: Map<string, number>;
+
+  private readonly decrementGuard: ReentrancyGuard;
+
+  constructor(maxQueue: number) {
+    this.maxQueue = maxQueue;
+    this.locks = new Map<string, number>();
+    this.decrementGuard = new ReentrancyGuard();
+  }
+
+  /**
+   * Increments the queue count for a particular origin
+   *
+   * @param origin
+   */
+  public increment(origin: string) {
+    const currentCount = this.locks.get(origin) ?? 0;
+    if (currentCount > this.maxQueue) {
+      throw new Error('Maximum number of requests reached. Try again later.');
+    }
+    this.locks.set(origin, currentCount + 1);
+  }
+
+  /**
+   * Decrements the queue count for a particular origin
+   *
+   * @param origin
+   */
+  public decrement(origin: string) {
+    if (this.decrementGuard.getDecrementLockStatus(origin)) {
```

Comment:
It's not clear to me why do you need a guard on decrement. (Also why not in increment?). Please explain in a comment why this is needed